src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,260 @@
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+typedef struct {
+    simplePid_s altPid;
+    simplePid_s velPid;
+    float throttle;
+    float targetAltitude;
+    float measuredAltitude;
+    float measuredAccel;
+    float velocityEstimationAccel;  // based on acceleration
+    float startVelocityEstimationAccel;
+    bool prevAltHoldModeEnabled;
+} altHoldState_s;
+
+void altHoldReset(altHoldState_s* altHoldState)
+{
+    simplePidInit(&altHoldState->altPid, -50.0f, 50.0f,
+                  0.01f * altholdConfig()->altPidP,
+                  0.0f,
+                  0.01f * altholdConfig()->altPidI);
+
+    simplePidInit(&altHoldState->velPid, 0.0f, 1.0f,
+                  0.01f * altholdConfig()->velPidP,
+                  0.01f * altholdConfig()->velPidD,
+                  0.0f);
+
+    altHoldState->throttle = 0.0f;
+    altHoldState->startVelocityEstimationAccel = altHoldState->velocityEstimationAccel;
+    altHoldState->targetAltitude = (float)(0.01f * getEstimatedAltitudeCm());
+}
+
+void altHoldInit(altHoldState_s* altHoldState)
+{
+    altHoldState->prevAltHoldModeEnabled = false;
+    altHoldState->velocityEstimationAccel = 0.0f;
+    altHoldReset(altHoldState);
+}
+
+
+// Rotate a vector *v by the euler angles defined by the 3-vector *delta.
+void rotateV(struct fp_vector *v, fp_angles_t *delta)
+{
+    struct fp_vector v_tmp = *v;
+
+    fp_rotationMatrix_t rotationMatrix;
+
+    buildRotationMatrix(delta, &rotationMatrix);
+
+    applyMatrixRotation((float *)&v_tmp, &rotationMatrix);
+
+    v->X = v_tmp.X;
+    v->Y = v_tmp.Y;
+    v->Z = v_tmp.Z;
+}
+
+void altHoldUpdateTarget(altHoldState_s* altHoldState)
+{
+    float rcThrottle = rcCommand[THROTTLE];
+
+    rcThrottle -= rxConfig()->midrc;
+    rcThrottle = constrainf(rcThrottle, -500.0f, 500.0f);
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, (int16_t)(rcThrottle));
+
+    rcThrottle = scaleRangef(rcThrottle, -500.0f, 500.0f, 0.0f, 1.0f);
+
+    DEBUG_SET(DEBUG_ALTHOLD, 1, (int16_t)(rcThrottle * 100.0f));
+
+    float altitudeChangeMaxSpeed = 0.1f * altholdConfig()->maxVerticalVelocity;
+
+    if (rcThrottle < 0.25f) {
+        rcThrottle = scaleRangef(rcThrottle, 0.0f, 0.25f, -1.0f, 0.0f);
+    } else if (rcThrottle > 0.75f) {
+        rcThrottle = scaleRangef(rcThrottle, 0.75f, 1.0f, 0.0f, 1.0f);
```

Comment:
this should be configurable and likely default to a much smaller window. I don't think people will want to move their throttle this much for an effect to occur. Should also look to deal with throttle curves and center throttle settings.

Reply:
For this mode people expect that the quad stay still if they do nothing. If we reduce window it will become harder to get into that "altitude hold" zone. May be it can be reduced a little bit but not too much.

Reply:
thats true sir